Introducing Kubernetes spec files auto-generator

aanm opened the pull request:
Introduced the Kubernetes spec files auto-generator.

The explanation can be found in `examples/kubernetes/README.rst` in this PR.

Some questions:
- Should we remove `examples/cilium.yaml`?
- How can we make use of directory structure in the documentation? (Example tabs like these ones maybe? http://cilium.readthedocs.io/en/latest/policy/language/?highlight=l7%20policy#simple-ingress-allow)
- How can we guarantee the generated files are being tested? cc @cilium/ci

aanm committed:
examples/kubernetes: add k8s spec file auto-generator

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
examples/kubernetes: add k8s spec files for master (:latest)

Signed-off-by: André Martins <andre@cilium.io>

tgraf reviewed (changes requested):
I don't think we should have Cilium version specific yamls in the code tree. We can access the tag specific yamls as we did before, e.g. `https://raw.githubusercontent.com/cilium/cilium/v1.0.0-rc7/examples/kubernetes/1.10/`

I think we should keep put the `ConfigMap` into a separate directory so we can `apply -f` the directory which contains all of the non ConfigMap resources and it will always update.

aanm committed:
docs/conf.py: Update copyright date to 2018

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
docs: add global var SCM_BRANCH for branch name

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
docs: update docs with tabs for multiple k8s versions

 - Minikube tutorial which uses a single `cilium.yaml` file
 - Upgrade which uses RBAC and cilium-ds.yaml file
 - Installation and QuickInstall which uses `cilium.yaml` and
   requires the user to modify the ConfigMap for etcd address

Signed-off-by: André Martins <andre@cilium.io>

raybejjani commented on Documentation/install/upgrade.rst:
```diff
@@ -9,14 +9,53 @@ Kubernetes Cilium Upgrade
+``cilium.yaml``, the safest way to upgrade Cilium to version "\ |SCM_BRANCH|" is
+by updating the RBAC rules and the DaemonSet file provided, which makes sure
+the ConfigMap already stored in the cluster will not be affected by the upgrade.
+Both files are dedicated to "\ |SCM_BRANCH|" for each Kubernetes version.
```
I found this paragraph a little confusing. Do we tell people to change the cilium-ds.yaml file somewhere? I would expect it to start with "The safest way to upgrade cilium is by updtaing the RBAC rules and Daemonset from the files provided. These make sure ..."

aanm replied:
Changed to:

>The safest way to upgrade Cilium to version "\ |SCM_BRANCH|" is by updating the
>RBAC rules and the DaemonSet file provided, which makes sure the ConfigMap,
>initially set up by ``cilium.yaml``, already stored in the cluster will not be
>affected by the upgrade.

raybejjani commented on Documentation/install/upgrade.rst:
```diff
@@ -9,14 +9,53 @@ Kubernetes Cilium Upgrade
+      $ kubectl apply -f \ |SCM_WEB|\/examples/kubernetes/1.11/cilium-ds.yaml
+
+
+You can also substitute the desired Cilium version number of vX.Y.Z in the
```
"Cilium version number _for_ vX.Y.Z"

aanm replied:
Done

raybejjani commented on Documentation/install/upgrade.rst:
```diff
@@ -9,14 +9,53 @@ Kubernetes Cilium Upgrade
+You can also substitute the desired Cilium version number of vX.Y.Z in the
+command below, but be aware the copy of the spec file stored in Kubernetes might
+run out-of-sync with the CLI flags, or options, specified by each Cilium
+version.
```
"but be aware _that_ the copy of the spec file stored in Kubernetes might _not be in-sync_"

aanm replied:
Done

raybejjani commented on examples/kubernetes/Makefile:
```diff
@@ -0,0 +1,41 @@
+	    cd $$k8s_version && \
+	    sed -f transforms2sed.sed ../templates/v1/$@ | \
+	    sed s+__CILIUM_VERSION__+$(CILIUM_VERSION)+g > "$*"; \
+	    cd ..; \
```
I'm not sure that it's any more clear, but if you run a subshell inside the loop (with `(blah)`) you won't need the `cd ..` and `cd -`.

aanm replied:
Cool, thanks!

raybejjani commented on examples/kubernetes/README.rst:
```diff
@@ -0,0 +1,58 @@
+specify the Cilium version yourself use :code:`make CILIUM_VERSION=X.Y.Z`.
+
+If you want to clean up a specific version, run :code:`make CILIUM_VERSION=X.Y.Z`
+which will delete all generated files for a specific Cilium version
```
How does this work? The makefile just clears things up and CILIUM_VERSION is used only to populate the image version, right?

aanm replied:
It's a leftover, previously I had a directory inside each k8s directory. Switched to

>If you want to clean up a specific version, run :code:`make clean` which will delete all generated files.

raybejjani replied:
cool. It would delete the copy checked out, right?

raybejjani commented on examples/kubernetes/README.rst:
```diff
@@ -0,0 +1,58 @@
+- :code:`cilium-sa.yaml` - The Cilium's Kubernetes :code:`ServiceAccount`.
+
+- :code:`cilium.yaml` - All previous files concatenated into a single file,
+  useful to deploy Cilium in a minikube environment with a "single line" command.
```
I guess we need this because we can't run `kubectl -f` on a url directory. Do we want to support running it on an on-disk copy? If so, we'd need to move this file out of the k8s version specific dir. Perhaps:
```
/1.8/cilium-{cm,ds,rbac,sa}.yaml
/cilium-1.8.yaml
```

It's probably not worth it, however, since it makes all our generators more complex.

aanm replied:
Doesn't that mean we need to have `/1.8/cilium-{cm,ds,rbac,sa}-1.8.yaml` as well?

raybejjani replied:
oh, my reasoning was that you can then do the `-f path/to/dir` if you have a local directory, but can still use `-f https://githubrawwhatever.com/cilium/cilium-1.8.yaml` for the tutorial examples. This cilium-1.8.yaml file is at the same level as the 1.8 directory.

raybejjani commented on examples/kubernetes/plugins/prometheus/prometheus.yaml:
```diff
@@ -0,0 +1,171 @@
+---
```
Could you also update the command references at http://cilium.readthedocs.io/en/latest/gettingstarted/minikube/#extra-metrics

aanm replied:
Done

raybejjani replied:
thanks 👍

aanm committed:
test: use generate k8s spec files for testing

- Since we added auto-generated files depending on which Kubernetes
version used, we need to test all of them accordingly to each specific
Kubernetes version.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
tests: disabling K8sValidatedUpdates test

Signed-off-by: André Martins <andre@cilium.io>

manalibhutiyani commented on test/k8sT/Updates.go:
```diff
@@ -11,7 +11,7 @@ import (
 )
 
-var _ = Describe("K8sValidatedUpdates", func() {
+var _ = Describe("DisabledK8sValidatedUpdates", func() {
```
Can you add a comment to #3442 to undo this?

aanm replied:
Done

manalibhutiyani replied:
There is also 
```
It("Updating Cilium stable to master", func() {
		// GH-3442 Re-enable once stable
		return
````
@aanm told me this was not working reliably and so he had to add this code. #3442 needs to undo both this `return` code as well as the `Disabled` keyword to enable this test correctly.

manalibhutiyani commented on examples/kubernetes/Makefile:
```diff
@@ -0,0 +1,38 @@
+
+clean:
+	for k8s_version in 1.7 1.8 1.9 1.10 1.11; do \
+        rm ./$$k8s_version/*.yaml; \
```
do we need a `cleanall` / `veryclean` to rm the directories as well created in .yaml and .sed?

manalibhutiyani replied:
You can ignore if not really required.

manalibhutiyani commented on examples/kubernetes/README.rst:
```diff
@@ -0,0 +1,58 @@
+The structure directory will be :code:`${k8s_major_version}.${k8s_minor_version}/*.yaml`.
+
+To generate those files simply run :code:`make`, which will pick up the Cilium
+version from the :code:`VERSION` file at the root of the repo, or if you want to
```
Nit: repo --> repository . Would be good to use more formal full words instead of abbreviations in the documentation.

aanm replied:
Done

manalibhutiyani commented on examples/kubernetes/README.rst:
```diff
@@ -0,0 +1,58 @@
+
+If you want to clean up a specific version, run :code:`make clean` which will
+delete all generated files.
+
```
If you decide to go ahead with `cleanall` or `veryclean` add comment about all directories being cleanup/removed up as well.

manalibhutiyani reviewed (approved):
Reviewed per commit. Looks good overall. Please find review comments per commit below.
